jenkinsci/plot-plugin
# Use @DataBoundSetter for optional pipeline params

This helps the Snippet Generator provide a smaller example
where many defaults are used, and prevents the Jenkins pipeline
linter from complaining about missing "required" parameters.

### JIRA

No issue directly found, though the closest I could find is [JENKINS-50636](https://issues.jenkins-ci.org/browse/JENKINS-50636).

### Misc

I started working on this due to build failures in my Jenkins pipeline when using a Jenkinsfile from SCM. This does not seem to happen with an inline pipeline job.

[plot-build-failure.txt](https://github.com/jenkinsci/plot-plugin/files/2046155/plot-build-failure.txt)

### What has been done
1. Optional constructor parameters have been removed and have been given `@DataBoundSetter`s
2. Generic `series` has been removed and csv, xml, properties are concatenated during `perform`.

### How to test
You can check if a Jenkinsfile will pass muster by using [the built-in linter](https://jenkins.io/doc/book/pipeline/development/#linter)

The following Jenkinsfile works with my changes, but fail with errors similar to the attached [plot-build-failure.txt](https://github.com/jenkinsci/plot-plugin/files/2046155/plot-build-failure.txt).

(You can also just create a simple Pipeline job and paste the following)
```groovy
pipeline {
  agent any
  stages {
    stage('write-data') {
      steps {
        sh '''
          echo '<my_data>' > data.xml
          echo '    <test value1="123.456"/>' >> data.xml
          echo '    <test value2="321.654"/>' >> data.xml
          echo '</my_data>' >> data.xml
        '''
      }
    }
  }
  post {
    always {
      plot csvFileName: 'plot-d519ae37-01b4-446b-a901-a98d0b28efe6.csv',
           group: 'My Data',
           title: 'Useful Title',
           style: 'line',
           yaxis: 'arbitrary',
           xmlSeries: [
             [file: 'data.xml',
              nodeType: 'NODESET',
              xpath: 'my_data/test/@*']
           ]
    }
  }
}
```

I tested this with Jenkins (LTS 2.107.3) inside Docker:
```
docker run -d --name jenkins -p 127.0.0.1:8080:8080 -p 50000:50000 -v jenkins_home:/var/jenkins_home jenkins/jenkins:lts
```

I installed the suggested plugins, and then uploaded the built hpi to test.


### Checklist
